### PR TITLE
.Net: bentho/local-runtime

### PR DESCRIPTION
**Pull request by alliscode**
### Description

Initial commit of local runtime for processes. Things coming in following PRs:

- Unit and Integration tests
- Support for state management
- Telemetry
- Much more.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        // Build the process to get a handle that can be started
+        KernelProcess kernelProcess = process.Build();
+
+        // Start the process with an initial external event
+        var runningProcess = await kernelProcess.StartAsync(kernel, new KernelProcessEvent() { Id = "StartProcess", Data = null });
+    }
+
+    public class IntroStep : KernelProcessStep
+    {
+        [KernelFunction()]
```
Comment (esttenorio) [inline dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs:64]: seems to work fine, but perhaps there should be a comment explaining that when there's 1 function per step no name is needed (?)

**Review by esttenorio — APPROVED** [review]

File: dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs (hunk 1 above)
Comment (alliscode) [reply dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs:64]: Oh this is an SK thing, when there is no name in the attribute it just used the function name.

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+#pragma warning disable CA1031 // Do not catch general exception types
+        try
+        {
+            invokeResult = await this.InvokeFunction(function, this._kernel, arguments).ConfigureAwait(false);
+            eventName = $"{targetFunction}.OnResult";
+            eventValue = invokeResult?.GetValue<object>();
+        }
+        catch (Exception ex)
+        {
+            this._logger?.LogError("Error in Step {StepName}: {ErrorMessage}", this.Name, ex.Message);
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:193]: Can we please consider filing an issue to introduce schematized logging messages.  Going to be a lot easier to start there than refactor a bunch of logging later.

https://learn.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+            eventName = $"{targetFunction}.OnError";
+            eventValue = ex.Message;
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:195]: I like how the failure is handled here.  I really got burned with the silent failure in the early poc.

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        // Activate the step with user-defined state if needed
+        KernelProcessStepState? stateObject = null;
+        Type? stateType = null;
+
+        if (TryGetSubtypeOfStatefulStep(stepInfo.InnerStepType, out Type? genericStepType) && genericStepType is not null)
+        {
+            // The step is a subclass of KernelProcessStep<>, so we need to extract the generic type argument
+            // and create an instance of the corresponding KernelProcessStepState<>.
+            var userStateType = genericStepType.GetGenericArguments()[0];
+            Verify.NotNull(userStateType);
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:84]: I don't believe this is proper use of `ArgumentException`

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+            stateType = typeof(KernelProcessStepState<>).MakeGenericType(userStateType);
+            Verify.NotNull(stateType);
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:87]: I don't believe this is proper use of `ArgumentException`

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+            stateObject = (KernelProcessStepState?)Activator.CreateInstance(stateType, this.Id, this.Name);
+        }
+        else
+        {
+            // The step is a KernelProcessStep with no user-defined state, so we can use the base KernelProcessStepState.
+            stateType = typeof(KernelProcessStepState);
+            stateObject = new KernelProcessStepState(this.Id, this.Name);
+        }
+
+        Verify.NotNull(stateObject);
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:98]: I don't believe this is proper use of `ArgumentException`

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        MethodInfo? methodInfo = stepInfo.InnerStepType.GetMethod(nameof(KernelProcessStep.ActivateAsync), [stateType]);
+        Verify.NotNull(methodInfo);
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:100]: I don't believe this is proper use of `ArgumentException`

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        }
+
+        return [];
+    }
+
+    public async Task HandleMessageAsync(LocalMessage message)
+    {
+        if (this._functions is null || this._inputs is null || this._initialInputs is null)
+        {
+            throw new InvalidOperationException("The step has not been initialized.");
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:132]: I might strongly consider introducing a `ProcessException` to that exceptions specific to this framework might be clearly distinguishable from other system classes.

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        MethodInfo? methodInfo = stepInfo.InnerStepType.GetMethod(nameof(KernelProcessStep.ActivateAsync), [stateType]);
+        Verify.NotNull(methodInfo);
+
+        methodInfo.Invoke(stepInstance, [stateObject]);
+        await stepInstance.ActivateAsync(stateObject).ConfigureAwait(false);
+    }
+
+    public IEnumerable<KernelProcessEvent> GetAllEvents()
+    {
+        var allEvents = this._eventQueue.ToList();
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:108]: Collection Nit - ToArray() is lighter and will suffice.

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+            {
+                this._inputs[message.FunctionName] = new();
+                functionParameters = this._inputs[message.FunctionName];
+            }
+
+            functionParameters![kvp.Key] = kvp.Value;
+        }
+
+        // If we're still waiting for inputs on all of our functions then don't do anything.
+        List<string> invocableFunctions = this._inputs.Where(i => i.Value != null && i.Value.All(v => v.Value != null)).Select(i => i.Key).ToList();
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:156]: Collection nit - `ToArray` will suffice:

`IList<string> invocableFunctions = ...ToArray();`

File: dotnet/src/Experimental/Process.LocalRuntime/LocalProcess.cs [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+            for (int superstep = 0; superstep < maxSupersteps; superstep++)
+            {
+                // Get all of the messages that have been sent to the steps within the process and queue them up for processing.
+                foreach (var step in this._steps)
+                {
+                    this.EnqueueStepMessagesAsync(step, localKernel, messageChannel);
+                }
+
+                // Complete the writing side, indicating no more messages in this superstep.
+                var messagesToProcess = messageChannel.ToList();
```
Comment (crickman) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalProcess.cs:119]: Collection nit - `ToArray` will suffice.

File: dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+            .SendEventTo(new ProcessFunctionTargetBuilder(userInputStep, "GetUserInput"));
+
+        // Build the process to get a handle that can be started
+        KernelProcess kernelProcess = process.Build();
+
+        // Start the process with an initial external event
+        var runningProcess = await kernelProcess.StartAsync(kernel, new KernelProcessEvent() { Id = "StartProcess", Data = null });
+    }
+
+    public class IntroStep : KernelProcessStep
```
Comment (markwallace-microsoft) [inline dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs:62]: Add XML documentation to all classes

File: dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+    {
+        [KernelFunction()]
+        public void PrintIntroMessage()
+        {
```
Comment (markwallace-microsoft) [inline dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs:66]: Should there be a Console write statement in here?

File: dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        }
+    }
+
+    public class UserInputStep : KernelProcessStep<UserInputState>
+    {
+        private UserInputState? _state;
```
Comment (markwallace-microsoft) [inline dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs:72]: Should `KernelProcessStepState` be stored in `KernelProcessStep`

Consider this pattern

```csharp
class KernelProcessStep
{
  KernelProcessStepState State { get; private set; } 
  
  // Default implementation store the state
  ValueTask ActivateAsync(KernelProcessStepState state)
  {
    this.State = state;
    await this.HandleActivatedAsync(state);
    return ValueTask.CompletedTask;
  }

  // Subclasses an optionally override this method to initialise the state.
  ValueTask HandleActivatedAsync(KernelProcessStepState state)
  {
  }
}
```

File: dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+                await context.EmitEventAsync(new() { Id = ChatBotEvents.Exit });
+                return;
+            }
+
+            // Emit the user input
+            await context.EmitEventAsync(new() { Id = ChatBotEvents.UserInputReceived, Data = input });
+        }
+    }
+
+    public class ChatBotResponseStep : KernelProcessStep<ChatBotState>
```
Comment (markwallace-microsoft) [inline dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs:104]: Can all of these classes be private sealed?

File: dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+
+        // Create a process that will interact with the chat completion service
+        ProcessBuilder process = new("ChatBot");
+        var introStep = process.AddStepFromType<IntroStep>();
+        var userInputStep = process.AddStepFromType<UserInputStep>();
+        var responseStep = process.AddStepFromType<ChatBotResponseStep>();
+
+        // Define the behavior when the process receives an external event
+        process.OnExternalEvent("StartProcess")
+            .SendEventTo(new ProcessFunctionTargetBuilder(introStep, "PrintIntroMessage"));
```
Comment (markwallace-microsoft) [inline dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs:33]: Is function name required when there is only a single function?

File: dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs [hunk 17]
```diff
@@ -0,0 +1,158 @@
...
+
+        // When the intro is complete, notify the userInput step
+        introStep
+            .OnFunctionResult("PrintIntroMessage")
+            .SendEventTo(new ProcessFunctionTargetBuilder(userInputStep, "GetUserInput"));
+
+        // When the userInput step emits an exit event, send it to the end steprt
```
Comment (markwallace-microsoft) [inline dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs:40]: Typo `steprt`

File: dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+        userInputStep
+            .OnFunctionResult("GetUserInput")
+            .StopProcess();
+
+        // When the userInput step emits a user input event, send it to the assistantResponse step
```
Comment (markwallace-microsoft) [inline dotnet/samples/GettingStartedWithProcesses/Step01_Processes.cs:45]: Should `assistantStep` be `responseStep`

File: dotnet/src/Experimental/Process.Core/Extensions/ProcessTypeExtensions.cs [hunk 19]
```diff
@@ -2,7 +2,7 @@
 
 using System;
 
-namespace Microsoft.SemanticKernel;
+namespace Microsoft.SemanticKernel.Process;
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.Core/Extensions/ProcessTypeExtensions.cs:5]: Will this be less discoverable now?

File: dotnet/src/Experimental/Process.Core/ProcessStepBuilder.cs [hunk 20]
```diff
@@ -3,6 +3,7 @@
 using System;
 using System.Collections.Generic;
 using System.Linq;
+using Microsoft.SemanticKernel.Process;
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.Core/ProcessStepBuilder.cs:6]: This seems odd, why is a subset of the functionality in a different namespace?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalProcessFactory.cs [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+/// A class that can run a process locally or in-process.
+/// </summary>
+public static class LocalKernelProcessFactory
+{
+    /// <summary>
+    /// Starts the specified process.
+    /// </summary>
+    /// <param name="process">Required: The <see cref="KernelProcess"/> to start running.</param>
+    /// <param name="kernel">Required: An instance of <see cref="Kernel"/></param>
+    /// <param name="initialEvent">Required: The initial event to start the process.</param>
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalProcessFactory.cs:19]: It says required here but the parameter is optional

File: dotnet/src/Experimental/Process.LocalRuntime/LocalProcessFactory.cs [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+    /// <returns>An instance of <see cref="KernelProcess"/> that can be used to interogate or stop the running process.</returns>
+    public static async Task<LocalKernelProcessContext> StartAsync(this KernelProcess process, Kernel kernel, KernelProcessEvent? initialEvent = null)
+    {
+        Verify.NotNull(process);
+        Verify.NotNullOrWhiteSpace(process.State?.Name);
+        Verify.NotNull(kernel);
+
+        var processContext = new LocalKernelProcessContext(process, kernel);
+        await processContext.StartWithEventAsync(initialEvent).ConfigureAwait(false);
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalProcessFactory.cs:28]: What happens if `initialEvent` is null

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+    private readonly Queue<KernelProcessEvent> _eventQueue;
+
+    protected readonly string? ParentProcessId;
+    protected readonly ILoggerFactory? LoggerFactory;
+    protected Dictionary<string, List<KernelProcessEdge>>? _outputEdges;
+
+    public readonly string Name;
+    public readonly string Id;
+
+    public LocalStep(string name, string id, Kernel kernel, string? parentProcessId = null, ILoggerFactory? loggerFactory = null)
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:38]: XML documentation

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 24]
```diff
@@ -0,0 +1,274 @@
...
+        Verify.NotNull(kernel);
+
+        this.Name = name;
+        this.Id = id;
+        this.ParentProcessId = parentProcessId;
+        this.LoggerFactory = loggerFactory;
+
+        this._kernel = kernel;
+        this._eventQueue = new();
+        this._functions = [];
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:51]: For these last can you just initialise the member variables directly?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 25]
```diff
@@ -0,0 +1,274 @@
...
+    }
+
+    public virtual async ValueTask InitializeAsync(KernelProcessStepInfo stepInfo)
+    {
+        this._logger = this.LoggerFactory?.CreateLogger(stepInfo.InnerStepType) ?? new NullLogger<LocalStep>();
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:56]: Should there be code to throw if this is called multiple times?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs [hunk 26]
```diff
@@ -0,0 +1,274 @@
...
+        // Instantiate an instance of the inner step object
+        KernelProcessStep stepInstance = (KernelProcessStep)ActivatorUtilities.CreateInstance(this._kernel.Services, stepInfo.InnerStepType);
+        var kernelPlugin = KernelPluginFactory.CreateFromObject(stepInstance);
+
+        // Load the kernel functions
+        foreach (KernelFunction f in kernelPlugin)
+        {
+            Verify.NotNull(f);
+            var cloned = f.Clone(stepInfo.State.Name!);
+            this._functions.Add(cloned.Name, cloned);
```
Comment (markwallace-microsoft) [inline dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:68]: Why is the cloning needed?

File: dotnet/src/Experimental/Process.LocalRuntime/LocalProcessFactory.cs (hunk 22 above)
Comment (alliscode) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalProcessFactory.cs:28]: Added parameter validation.

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs (hunk 26 above)
Comment (alliscode) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:68]: Good catch, this is not needed anymore. Updated it to set the plugin name when loading the plugin from the factory.

File: dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs (hunk 8 above)
Comment (alliscode) [reply dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs:132]: I'll use KernelException for now and create a work item to determine if a process specific exception is useful.

File: dotnet/src/Experimental/Process.Abstractions/KernelProcessStepInfo.cs [hunk 27]
```diff
@@ -19,32 +19,18 @@ public class KernelProcessStepInfo
...
-    /// A read-only collection of event Ids that this Step can emit.
-    /// </summary>
-    public IReadOnlyCollection<string> EventIds => this._outputEdges.Keys.ToArray();
+    public Type InnerStepType { get; }
 
     /// <summary>
     /// The state of the Step.
     /// </summary>
-    public KernelProcessStepState State { get; }
+    public KernelProcessStepState State { get; init; }
```
[the remark this answers is not in the document]
Comment (alliscode) [reply dotnet/src/Experimental/Process.Abstractions/KernelProcessStepInfo.cs:27]: This one is an interesting case. The goal is for the `KernelProcessStepInfo` to be immutable and work with the spread operator to make changes to the state when needed as in:

`kernelStep = kernelStep with { State = kernelStep.State with { Id = Guid.NewGuid().ToString() } };`

It appears that we do not support the `required` keyword due to support for dotnet standard 2.0 which makes this challenging to do cleanly. For now I'm keeping the pattern as is but adding validation to the init setter.

**Review by markwallace-microsoft — APPROVED** [review]

**Pull request merged**